pkg/trace/local_client.go

```diff
@@ -0,0 +1,145 @@
+	// extra buffer to avoid blocking the caller when writing to files.
+	cannal chan Event
+}
+
+// NewLocalClient creates a struct that will save all of the events passed to
+// the retuen channel to files based on their "table" (a string field in the
+// event). Each type gets its own file. The internal are purposefully not thread
+// safe to avoid the overhead of locking with each event save. Only pass events
+// to the returned channel. Call CloseAll to close all open files. Goroutine to
+// save events is started in this function.
+func NewLocalClient(cfg *config.InstrumentationConfig, logger log.Logger, chainID, nodeID string) (*LocalClient, error) {
+	fm := make(map[string]*os.File)
+	for _, table := range splitAndTrimEmpty(cfg.TracingTables, ",", " ") {
+		fileName := fmt.Sprintf("%s.jsonl", table)
+		file, err := os.OpenFile(fileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+		if err != nil {
+			return nil, fmt.Errorf("failed to open or create file %s: %w", fileName, err)
+		}
+		fm[table] = file
+	}
+	lc := &LocalClient{
+		fileMap: fm,
+		cannal:  make(chan Event, cfg.TraceBufferSize),
+		chainID: chainID,
+		nodeID:  nodeID,
+		logger:  logger,
+		mu:      sync.RWMutex{},
+	}
+	go lc.drainCannal()
+	return lc, nil
+}
+
+func (fm *LocalClient) Write(table string, data map[string]interface{}) {
+	if !fm.IsCollecting(table) {
+		return
+	}
+	fm.cannal <- NewEvent(fm.chainID, fm.nodeID, table, data)
+}
+
+func (fm *LocalClient) ReadTable(table string) ([]byte, error) {
+	return nil, errors.New("reading not supported using the Local tracing client")
+}
+
+func (fm *LocalClient) IsCollecting(table string) bool {
+	if _, has := fm.getFile(table); has {
+		return true
+	}
+	return false
+}
+
+// getFile gets or creates a file for the given type. This method is purposely
+// not thread-safe to avoid the overhead of locking with each event save.
+func (fm *LocalClient) getFile(table string) (*os.File, bool) {
+	f, has := fm.fileMap[table]
+	return f, has
+}
+
+// saveEventToFile marshals an Event into JSON and appends it to a file named after the event's Type.
+func (fm *LocalClient) saveEventToFile(event Event) error {
+	file, has := fm.getFile(event.Table)
+	if !has {
+		return fmt.Errorf("table %s not found", event.Table)
+	}
+
+	eventJSON, err := json.Marshal(event)
+	if err != nil {
+		return fmt.Errorf("failed to marshal event: %v", err)
+	}
+
+	if _, err := file.Write(append(eventJSON, '\n')); err != nil {
+		return fmt.Errorf("failed to write event to file: %v", err)
+	}
+
+	return nil
+}
+
+// drainCannal takes a variadic number of channels of Event pointers and drains them into files.
+func (fm *LocalClient) drainCannal() {
+	// purposefully do not lock, and rely on the channel to provide sync
+	// actions, to avoid overhead of locking with each event save.
+	for ev := range fm.cannal {
+		if err := fm.saveEventToFile(ev); err != nil {
+			fm.logger.Error("failed to save event to file", "error", err)
+		}
+	}
+}
```

Comment:
by not locking here, we get a lot closer to performance of having individual channels. If we wanted to get a bit more clever, we could additionally split into different channels which serialize and write to files in parallel.

I might do this, along with not blocking when we write to ensure that we never mess with performance. although we might want to keep track of when that occurs.